posts/2022-10-19-runtime-multistage-programming.adoc

```diff
@@ -0,0 +1,315 @@
+# Custom JIT compilation with Runtime Multi-Stage Programming
+
+For a while now I've been fiddling with a library called Slinc, a way to link C code in Scala. Since my original version, I've been using context values, macros, and other tools to generate code for writing and reading representations of Structs into and out of native memory.
+
+If you're not familiar with how macros work in Scala, the most simple way of writing them involves using quotes and splices.
+
+```scala
+val additionExpr = '{4 + 3}
+val function = '{() => $additionExpr} //produces () => 4+3
+```
+
+Compile time information can be used in combination with creative usage of quotes and splices to generate all kinds of different code:
+
+```scala
+val exprs = for 
+  i <- 0 until 5
+yield '{(j: Int) => println(j + ${Expr(i)})} //${Expr(i)} smuggles the compile time value i into our generated code
+
+'{
+	val k = 15
+	
+	${
+		Expr.block(
+			//would normally produce ((j: Int) => println(j + 0))(k) as an example
+			//but Expr.betaReduce simplifies it to println(k + 0)
+			exprs.map(expr => Expr.betaReduce('{$expr('k)})),
+			'{}
+		)
+	}
+}
+//resulting code:
+//{
+//  val k = 15
+//  println(k + 0)
+//  println(k + 1)
+//  println(k + 2)
+//  println(k + 3)
+//  println(k + 4)
+//}
+```
+
+Compile-time reflection and terms are also a thing in Scala 3 macros, but the usage of those is a story for another time.
+
+Anyway one of my first attempts at generating code to write data to native memory would take case classes like these:
```

Comment:
After reading through the article and then starting this review, my impression is that it jumps right into the deep end. I think it would benefit from having a section in the beginning of the text to introduce the concepts/model of the Slinc library.
For example; `Struct` is used out of the blue without any context

Reply:
The focus of the article isn't really Slinc per-se, but rather the runtime multi-stage programming feature of Scala 3, and the benefits it brings. The benefit outlined is that compile time code generation cannot create perfect code that writes a case class to native memory, because the offsets that the host environment would need to use is information only available at runtime, and that runtime multi-stage programming, being able to generate code based off of runtime information, solves this limitation of macros.

Reply:
Basically, Slinc and its needs (copying data into native memory), are in the article only to provide a real-world example of the applicability of this functionality.

Reply:
> The focus of the article isn't really Slinc per-se, but rather the runtime multi-stage programming feature of Scala 3, and the benefits it brings. 

Understood, but it would be a good idea nevertheless to point that out and give some more background.

Reply:
Ok I'll try to do that

Reply:
I think the updated version I wrote works nicely